packages/cli/command_spec.pony

```diff
@@ -0,0 +1,278 @@
+"""
+This package implements command line parsing with the notion of commands that
+are specified as a hierarchy.
+See RFC-xxx for more details.
+
+The general EBNF of the command line looks like:
+  command_line ::= root_command (option | command)* (option | arg)*
+  command ::= alphanum_word
+  alphanum_word ::= alphachar(alphachar | numchar | '_' | '-')*
+  option ::= longoption | shortoptionset
+  longoption ::= '--'alphanum_word['='arg | ' 'arg]
+  shortoptionset := '-'alphachar[alphachar]...['='arg | ' 'arg]
+  arg := boolarg | intarg | floatarg | stringarg
+  boolarg := 'true' | 'false'
+  intarg> := ['-'] numchar...
+  floatarg ::= ['-'] numchar... ['.' numchar...]
+  stringarg ::= anychar
+
+Some Examples:
+  usage: chat [<options>] <command> [<options>] [<args> ...]
+"""
+use col = "collections"
+
+class CommandSpec
+  """
+  CommandSpec describes the specification of a parent or leaf command. Each
+  command has the following attributes:
+
+  - a name: a simple string token that identifies the command.
+  - a description: used in the syntax message.
+  - a map of options: the valid options for this command.
+  - an optional help option+command name for help parsing
+  - one of:
+     - a Map of child commands.
+     - an Array of arguments.
+  """
+  let name: String
+  let descr: String
+  let options: col.Map[String, OptionSpec] = options.create()
+  var help_name: String = "xxx" // TODO: make this nice.
```

Comment:
I will remove it before merging. I introduced it with some last minute help improvements, so I'll double check things and remove it.